formal-spec/Leios/Spec.agda

```diff
@@ -0,0 +1,255 @@
+        BaseState : B.State
+
+  lookupEB : EBRef → Maybe EndorserBlock
+  lookupEB r = find (λ b → getEBRef b ≟ r) EBs
+
+  lookupIB : IBRef → Maybe InputBlock
+  lookupIB r = find (λ b → getIBRef b ≟ r) IBs
+
+  lookupTxs : EndorserBlock → List Tx
+  lookupTxs eb = do
+    eb′ ← mapMaybe lookupEB $ ebRefs eb
+    ib  ← mapMaybe lookupIB $ ibRefs eb′
+    txs $ body ib
+    where open EndorserBlockOSig
+          open IBBody
+          open InputBlock
+
+  constructLedger : List EndorserBlock → List Tx
+  constructLedger = concatMap lookupTxs
+
+  needsUpkeep : SlotUpkeep → Set
+  needsUpkeep = _∉ Upkeep
+
+addUpkeep : LeiosState → SlotUpkeep → LeiosState
+addUpkeep s u = let open LeiosState s in record s { Upkeep = Upkeep ∪ ❴ u ❵ }
+
+initLeiosState : VTy → StakeDistr → B.State → LeiosState
+initLeiosState V SD bs = record
+  { V         = V
+  ; SD        = SD
+  ; FFDState  = FFD.initFFDState
+  ; Ledger    = []
+  ; ToPropose = []
+  ; IBs       = []
+  ; EBs       = []
+  ; Vs        = []
+  ; slot      = initSlot V
+  ; IBHeaders = []
+  ; IBBodies  = []
+  ; Upkeep    = ∅
+  ; BaseState = bs
+  }
+
+-- some predicates about EBs
+module _ (s : LeiosState) (eb : EndorserBlock) where
+  open EndorserBlockOSig eb
+  open LeiosState s
+
+  allIBRefsKnown : Type
+  allIBRefsKnown = ∀[ ref ∈ fromList ibRefs ] ref ∈ˡ map getIBRef IBs
+
+stake : LeiosState → ℕ
+stake record { SD = SD } = case lookupᵐ? SD id of λ where
+  (just s) → s
+  nothing  → 0
+
+module _ (s : LeiosState) where
+
+  open LeiosState s
+
+  upd : Header ⊎ Body → LeiosState
+  upd (inj₁ (ebHeader eb)) = record s { EBs = eb ∷ EBs }
+  upd (inj₁ (vHeader vs)) = record s { Vs = vs ∷ Vs }
+  upd (inj₁ (ibHeader h)) with A.any? (matchIB? h) IBBodies
+  ... | yes p =
+    record s
+      { IBs = record { header = h ; body = A.lookup p } ∷ IBs
+      ; IBBodies = IBBodies A.─ p
+      }
+  ... | no _ =
+    record s
+      { IBHeaders = h ∷ IBHeaders
+      }
+  upd (inj₂ (ibBody b)) with A.any? (flip matchIB? b) IBHeaders
+  ... | yes p =
+    record s
+      { IBs = record { header = A.lookup p ; body = b } ∷ IBs
+      ; IBHeaders = IBHeaders A.─ p
+      }
+  ... | no _ =
+    record s
+      { IBBodies = b ∷ IBBodies
+      }
+
+infix 25 _↑_
+_↑_ : LeiosState → List (Header ⊎ Body) → LeiosState
+_↑_ = foldr (flip upd)
+
+open FFD hiding (_-⟦_/_⟧⇀_)
+
+private variable s s'   : LeiosState
+                 ffds'  : FFD.State
+                 bs'    : B.State
+                 ks ks' : K.State
+                 msgs   : List (FFDAbstract.Header ffdAbstract ⊎ FFDAbstract.Body ffdAbstract)
+                 eb     : EndorserBlock
+                 ebs    : List EndorserBlock
+                 txs    : List Tx
+                 V      : VTy
+                 SD     : StakeDistr
+                 pks    : List PubKey
+
+module Protocol (_⇀_ : LeiosState → LeiosState → Type) (allUpkeep : ℙ SlotUpkeep) (open Leios.Voting a) (va : VotingAbstract LeiosState) (open VotingAbstract va)
```

Comment:
I'm not sure if I like `⇀`. In the Ledger we're using this notation for maps, so it looks a lot like that to me. But it does make sense with the other arrows we have, and I don't see any other obvious candidate while staring at the options in the Agda input method.